### PR TITLE
Feature rules2dm

### PR DESCRIPTION
**Pull request by nikokaoja**
### Added

- Generation of data model in DMS through `sdk` interaction with DMS endpoint

### CODE REVIEW

File: cognite/neat/rules/_exceptions.py [hunk 1]
```diff
@@ -758,6 +758,112 @@ def __init__(self, field_name: str = "", no_of_values: int = None, verbose=False
...
+            "Only DatatypeProperty and ObjectProperty are supported!"
+        )
+        if verbose:
+            self.message += f"\nDescription: {self.description}"
+            self.message += f"\nExample: {self.example}"
+            self.message += f"\nFix: {self.fix}"
+
+
+class Warning61(NeatWarning):
+    type_: str = "ContainerPropertyTypeUnsupported"
```
Comment (doctrino) [inline cognite/neat/rules/_exceptions.py:784]: Copy paste error?
```suggestion
    type_: str = "ViewPropertyTypeUnsupported"
```

File: cognite/neat/rules/_validation.py [hunk 2]
```diff
@@ -1,10 +1,23 @@
...
+@overload
+def are_entity_names_dms_compliant(
+    transformation_rules: TransformationRules, return_report: bool = False
+) -> tuple[bool, list[dict]]:
+    ...
+
+
+@overload
+def are_entity_names_dms_compliant(transformation_rules: TransformationRules, return_report: bool = False) -> bool:
+    ...
```
Comment (doctrino) [inline cognite/neat/rules/_validation.py:18]: The same signature with different return value. 

```suggestion
@overload
def are_entity_names_dms_compliant(
    transformation_rules: TransformationRules, return_report: Literal[True]
) -> tuple[bool, list[dict]]:
    ...


@overload
def are_entity_names_dms_compliant(transformation_rules: TransformationRules, return_report: Literal[False] = False) -> bool:
    ...
```

I know this a strange syntax with `Literal[False]` but it is just the way it is.

File: cognite/neat/rules/_validation.py [hunk 3]
```diff
@@ -52,6 +65,18 @@ def are_entity_names_dms_compliant(
...
+@overload
+def are_properties_redefined(
+    transformation_rules: TransformationRules, return_report: bool = False
+) -> tuple[bool, list[dict]]:
+    ...
+
+
+@overload
+def are_properties_redefined(transformation_rules: TransformationRules, return_report: bool = False) -> bool:
+    ...
```
Comment (doctrino) [inline cognite/neat/rules/_validation.py:77]: Same issue here.

File: tests/rules/exporter/test_rules2dms.py [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+from cognite.neat.rules._exceptions import Error10
+
+
+def test_rules2dms(simple_rules):
+    data_model = DataModel.from_rules(transformation_rules=simple_rules)
+
+    assert len(data_model.containers) == 4
+    assert len(data_model.views) == 4
+    assert list(data_model.views.keys()) == ["CountryGroup", "Country", "PriceArea", "PriceAreaConnection"]
+    assert list(data_model.containers.keys()) == ["CountryGroup", "Country", "PriceArea", "PriceAreaConnection"]
```
Comment (doctrino) [inline tests/rules/exporter/test_rules2dms.py:12]:
```suggestion
    assert list(data_model.views) == ["CountryGroup", "Country", "PriceArea", "PriceAreaConnection"]
    assert list(data_model.containers) == ["CountryGroup", "Country", "PriceArea", "PriceAreaConnection"]
```

Keys are default for dicts

File: cognite/neat/rules/exporter/dm/rules2dms.py [hunk 5]
```diff
@@ -1,2 +1,315 @@
-"""Exports rules to CDF Data Model Storage.
```
Comment (doctrino) [inline cognite/neat/rules/exporter/dm/rules2dms.py:1]: Why is this in a submodule `dm` . Suggest moving it up one level as it is only one file.

File: cognite/neat/rules/exporter/dm/rules2dms.py [hunk 6]
```diff
@@ -1,2 +1,315 @@
...
+    name: Optional[str] = None
+    containers: dict[str, ContainerApply]
+    views: dict[str, ViewApply]
+
+    model_config: ClassVar[ConfigDict] = ConfigDict(
+        populate_by_name=True, str_strip_whitespace=True, arbitrary_types_allowed=True, strict=False, extra="allow"
+    )
+
+    @classmethod
+    def from_rules(cls, transformation_rules: TransformationRules, fix_names: bool = False) -> Self:
```
Comment (doctrino) [inline cognite/neat/rules/exporter/dm/rules2dms.py:51]: `fix_names` is not used.

File: cognite/neat/rules/exporter/dm/rules2dms.py [hunk 7]
```diff
@@ -1,2 +1,315 @@
...
+                _exceptions.Warning62(existing_containers, self.space).message,
+                category=_exceptions.Warning62,
+                stacklevel=2,
+            )
+        else:
+            for container_id in self.containers:
+                logging.info(f"Creating container {container_id} in space {self.space}")
+                _ = client.data_modeling.containers.apply(self.containers[container_id])
+
+    def create_views(self, client: CogniteClient, force: bool = False):
```
Comment (doctrino) [inline cognite/neat/rules/exporter/dm/rules2dms.py:251]: `force` is not used

File: cognite/neat/rules/exporter/dm/rules2dms.py [hunk 8]
```diff
@@ -1,2 +1,315 @@
...
+        cdf_views = {}
+        if views := client.data_modeling.views.list(space=self.space, limit=-1):
+            cdf_views = {container.external_id: container for container in views}
```
Comment (doctrino) [inline cognite/neat/rules/exporter/dm/rules2dms.py:254]:
```suggestion
            cdf_views = {view.external_id: view for view in views}
```

File: cognite/neat/rules/exporter/dm/rules2dms.py [hunk 9]
```diff
@@ -1,2 +1,315 @@
...
+            logging.warning(_exceptions.Warning63(existing_views, self.version, self.space).message)
+            warnings.warn(
+                _exceptions.Warning63(existing_views, self.version, self.space).message,
+                category=_exceptions.Warning63,
+                stacklevel=2,
+            )
+        else:
+            for view_id in self.views:
+                logging.info(f"Creating view {view_id} version {self.views[view_id].version} in space {self.space}")
+                _ = client.data_modeling.views.apply(self.views[view_id])
```
Comment (doctrino) [inline cognite/neat/rules/exporter/dm/rules2dms.py:266]: This looks strange to me. Why are you not creating the views when you get a warning?

File: cognite/neat/rules/exporter/dm/rules2dms.py [hunk 10]
```diff
@@ -1,2 +1,315 @@
...
+            _ = client.data_modeling.data_models.apply(
+                DataModelApply(
+                    name=self.name,
+                    description=self.description,
+                    space=self.space,
+                    external_id=self.external_id,
+                    version=self.version,
+                    views=list(self.views.values()),
+                )
+            )
```
Comment (doctrino) [inline cognite/neat/rules/exporter/dm/rules2dms.py:292]: Same here

**Review by doctrino — COMMENTED** [review]
Nice to see once you have the right classes in `DMS` then the transformation rules maps nicely to data models in CDF.

File: cognite/neat/rules/_exceptions.py (hunk 1 above)
Comment (nikokaoja) [reply cognite/neat/rules/_exceptions.py:784]: Fixed

File: cognite/neat/rules/exporter/dm/rules2dms.py (hunk 5 above)
Comment (nikokaoja) [reply cognite/neat/rules/exporter/dm/rules2dms.py:1]: moved on step up, it was purely for org purpose

File: cognite/neat/rules/exporter/dm/rules2dms.py (hunk 9 above)
Comment (nikokaoja) [reply cognite/neat/rules/exporter/dm/rules2dms.py]: I moved test if any of data model components exsist and if they do... then raise `Error60`

**Pull request merged**